lang/transformations/src/lifting/fv.rs

```diff
@@ -40,211 +41,236 @@ use ast::{Hole, TypeUniv, Variable};
 /// - `free_vars_closure(arg, ctx) ⊆ ctx`
 ///
 pub fn free_vars_closure<T: FV>(arg: &T, ctx: &TypeCtx) -> FreeVars {
-    let mut v = FreeVarsVisitor { fvs: Default::default(), type_ctx: ctx, lvl_ctx: ctx.levels() };
-
-    arg.visit_fv(&mut v);
-
-    FreeVars { fvs: v.fvs }
+    // Start with the context’s level context and the given type context.
+    arg.free_vars_closure(ctx.levels().clone(), ctx.clone())
 }
 
-/// Vistor trait for computing the closure of free variables closed under type dependencies
-///
-/// This vistor implements an AST traversal that stores any free variables it encounters.
-/// It uses `FreeVarsVisitor` to store the free variables and to keep track of the context.
-/// The interesting case is when a variable is encountered: `impl FV for Variable`.
+/// A trait for computing the free variables of an AST node (transitive closure).
 pub trait FV {
-    fn visit_fv(&self, v: &mut FreeVarsVisitor);
+    /// Returns the closure of free variables in `self` that are bound in `type_ctx`.
+    ///
+    /// # Parameters
+    ///
+    /// - `lvl_ctx`:  The context that contains all free variables in `self`.
+    /// - `type_ctx`: The context that contains the types of free variables of interest.
+    ///
+    /// # Requires
+    ///
+    /// - `self` is well-typed in `lvl_ctx`.
+    /// - `type_ctx ⊆ lvl_ctx`.
+    ///
+    /// # Returns
+    ///
+    /// The closure of free variables in `self` closed under type dependencies with respect to `type_ctx`.
+    ///
+    /// # Ensures
+    ///
+    /// - `free_vars_closure(self, lvl_ctx, type_ctx) ⊆ type_ctx`
+    fn free_vars_closure(&self, lvl_ctx: LevelCtx, type_ctx: TypeCtx) -> FreeVars;
 }
 
-impl FV for Vec<Case> {
-    fn visit_fv(&self, v: &mut FreeVarsVisitor) {
-        for case in self {
-            case.visit_fv(v)
-        }
+/// A set of free variables.
```

Comment:
Do we need `FreeVars`  as a struct? We could just reuse `HashSet<FreeVar>`, since the hashset does not come with any interesting invariants and only forwards the implementation of insert and extend.

Reply:
`subst.rs` still requires `FreeVars` to be a newtype but this can be resolved in the next PR.